File: social-service/src/main/java/com/eirsteir/coffeewithme/social/config/UserConfiguration.java
package com.eirsteir.coffeewithme.social.config;

import com.eirsteir.coffeewithme.social.repository.FriendshipRepository;
import com.eirsteir.coffeewithme.social.repository.UniversityRepository;
import com.eirsteir.coffeewithme.social.repository.UserRepository;
import com.eirsteir.coffeewithme.social.service.AccountEventConsumer;
import com.eirsteir.coffeewithme.social.service.user.UserService;
import com.eirsteir.coffeewithme.social.service.user.UserServiceImpl;
import io.eventuate.tram.events.subscriber.DomainEventDispatcher;
import io.eventuate.tram.events.subscriber.DomainEventDispatcherFactory;
import io.eventuate.tram.spring.events.publisher.TramEventsPublisherConfiguration;
import io.eventuate.tram.spring.events.subscriber.TramEventSubscriberConfiguration;
import io.eventuate.tram.spring.jdbckafka.TramJdbcKafkaConfiguration;
import io.eventuate.tram.spring.optimisticlocking.OptimisticLockingDecoratorConfiguration;
import org.modelmapper.ModelMapper;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;

@Configuration
public class UserConfiguration {

  @Bean
  public AccountEventConsumer accountEventConsumer() {
    return new AccountEventConsumer();
  }

  @Bean
  public UserService userService(
      UserRepository userRepository,
      UniversityRepository universityRepository,
      FriendshipRepository friendshipRepository,
      ModelMapper modelMapper) {
    return new UserServiceImpl(
        userRepository, universityRepository, friendshipRepository, modelMapper);
  }

  @Bean
  public DomainEventDispatcher domainEventDispatcher(
      AccountEventConsumer accountEventConsumer,
      DomainEventDispatcherFactory domainEventDispatcherFactory) {
    return domainEventDispatcherFactory.make(
        "accountServiceEvents", accountEventConsumer.domainEventHandlers());
  }
}

File: social-service/src/main/kotlin/com/eirsteir/coffeewithme/social/config/UserConfiguration.kt
package com.eirsteir.coffeewithme.social.config

import com.eirsteir.coffeewithme.social.repository.FriendshipRepository
import com.eirsteir.coffeewithme.social.repository.UniversityRepository
import com.eirsteir.coffeewithme.social.repository.UserRepository
import com.eirsteir.coffeewithme.social.service.AccountEventConsumer
import com.eirsteir.coffeewithme.social.service.user.UserService
import com.eirsteir.coffeewithme.social.service.user.UserServiceImpl
import io.eventuate.tram.events.publisher.DomainEventPublisher
import io.eventuate.tram.events.subscriber.DomainEventDispatcher
import io.eventuate.tram.events.subscriber.DomainEventDispatcherFactory
import org.modelmapper.ModelMapper
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class UserConfiguration {
    @Bean
    fun accountEventConsumer(
        domainEventPublisher: DomainEventPublisher,
        userRepository: UserRepository): AccountEventConsumer =
        AccountEventConsumer(userRepository, domainEventPublisher)

    @Bean
    fun userService(
        userRepository: UserRepository,
        universityRepository: UniversityRepository,
        friendshipRepository: FriendshipRepository,
        modelMapper: ModelMapper
    ): UserService = UserServiceImpl(
        userRepository, universityRepository, friendshipRepository, modelMapper
    )

    @Bean
    fun domainEventDispatcher(
        accountEventConsumer: AccountEventConsumer,
        domainEventDispatcherFactory: DomainEventDispatcherFactory
    ): DomainEventDispatcher = domainEventDispatcherFactory.make(
        "accountServiceEvents", accountEventConsumer.domainEventHandlers()
    )
}